### PR TITLE
Base Implementation

### PR DESCRIPTION
**Pull request by tanimutomo**
base implementation

残りタスク
- [x] Implement Model
- [x] Implement Dataset
- [x] Implement Loss
- [x] Implement Train
- [x] Implement Test

- [x] Test run model.py
- [x] Test run dataset.py
- [x] Test run loss.py
- [x] Test run train.py
- [x] Test run test.py

- [x] Finish downloading the dataset in chasca

### CODE REVIEW

**Comment by tanimutomo** [comment]
PartialConvを実装してU-Netモデルも実装した
本家実装と，inoueさんの実装を参考にした

**Comment by tanimutomo** [comment]
lossはinoueさんの実装を参考にして，
https://github.com/naoto0804/pytorch-inpainting-with-partial-conv/blob/master/loss.py
random mask creationはkeras実装の方を参考にする
https://github.com/MathiasGruber/PConv-Keras

**Comment by tanimutomo** [comment]
- inoueさんのtv_lossの実装でtorch.meanをしているが，する必要性がない気がするので，pytorochのl1lossに突っ込むように実装を変更する
- また，l1lossはtorch.absでも実装可能だが，pytorchのnn.L1Lossに入れるとCのextensionで計算してくれるので，そっちを使った方が計算効率が圧倒的に高い気がする

**Comment by tanimutomo** [comment]
places2の512のデータだと大きすぎる問題と，学習が終わらない問題が発生しそうだったので，論文実装用に256の画像で学習するように方向転換した．
そのためにネットワークの深さをlayer_sizeで自由に調整可能にした
また，implement valid and hole loss.

**Comment by tanimutomo** [comment]
pytorchのdatasetとdataloaderのtutorialを使って，作り方を見直した
dataset normalizationのmeanとstdは直接計算した値をそのまま使用していると思われる．（いちいち計算するのが面倒なので，一度計算した値を使いまわしている）

**Comment by tanimutomo** [comment]
inoueさんのやつとpytorch tutorialを参考にして，Places datasetクラスを作成

**Comment by tanimutomo** [comment]
trainingやmodelのparameterを指定するためのconfig.pyを追加

**Comment by tanimutomo** [comment]
以下がVGG16 model overview
```
VGG(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): ReLU(inplace)
    (2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (3): ReLU(inplace)
    (4): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (5): Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (6): ReLU(inplace)
    (7): Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (8): ReLU(inplace)
    (9): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (10): Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): ReLU(inplace)
    (12): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (13): ReLU(inplace)
    (14): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (15): ReLU(inplace)
    (16): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (17): Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (18): ReLU(inplace)
    (19): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (20): ReLU(inplace)
    (21): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (22): ReLU(inplace)
    (23): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
    (24): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (25): ReLU(inplace)
    (26): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (27): ReLU(inplace)
    (28): Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (29): ReLU(inplace)
    (30): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(7, 7))
  (classifier): Sequential(
    (0): Linear(in_features=25088, out_features=4096, bias=True)
    (1): ReLU(inplace)
    (2): Dropout(p=0.5)
    (3): Linear(in_features=4096, out_features=4096, bias=True)
    (4): ReLU(inplace)
    (5): Dropout(p=0.5)
    (6): Linear(in_features=4096, out_features=1000, bias=True)
  )
)
```
partial conv inpaintingのperceptual lossでは，pool1, pool2, pool3のfeature mapを利用して計算する．

そのため，ネットワークを4, 9, 16のfeature mapを出力できるようにsequentialで区切る
```
vgg16 = models.vgg16(pretrained=True)
self.enc_1 = nn.Sequential(*vgg16.features[:5])
self.enc_2 = nn.Sequential(*vgg16.features[5:10])
self.enc_3 = nn.Sequential(*vgg16.features[10:17])
```
これでそれぞれのencのoutputを返せば，各地点でのfeature mapをgetすることができる

**Comment by tanimutomo** [comment]
また，vggのparameterはrequired_grad = Falseにして，計算グラフを作成しないようにする

**Comment by tanimutomo** [comment]
- configのバグを直して，Lossの実装を終えた．
- 基本的にはInoueさんのやつを参考にしたが，TVLossのところだけ実装が間違っていたので，そこは修正した．
- test runも動いた

**Comment by tanimutomo** [comment]
inoueさんの実装では，data transformationで，ToTensorで[0.0, 1.0]にしたあとに，VGGの入力として求めレているMEANとSTDのnormalizationを行なっていた．
しかし，このVGGの入力として求められている正規化は，果たしてここでやる必要があるのだろうか？
PConvUNetの出力として得られた画像に対して，このNormalizationを適用してVGGに流してLossを計算すればいいと思う．
そのため，この実装ではNormalizationはVGGのところで適用する

**Comment by tanimutomo** [comment]
EdgeConnectもStyle LossとPerceptual Lossを使っているので，調べたが，
こっちはVGGの入力前も最初のDatasetのTransformのところでも，ToTensorのNormalization以外に，正規化的なことをしているところはないし，MEANやSTDも特に定義されてないので，もしかしたらもはやVGGの正規化をしてない可能性もある．
そのため，以下の場合について精度検証をする
- VGGでLossを計算する直前にz-score normalizationをする
- PConvUNetへの入力の直前で（つまり，Data transformの段階で）やる
- 適用しない

**Comment by tanimutomo** [comment]
- mainとtrainを一旦実装した．
- とりあえずcomet-mlに頼って無駄なログや画像は残さないように実装した
- lossに関しては，logのためにそれぞれのlossを返すdictをreturnするように変更した．
- model initializerはとりあえずxavierを使う

**Comment by tanimutomo** [comment]
論文を読んでたら増えたタスク
- [ ] trainとtestでは当然異なるマスクを使用する
- [x] 使用するマスクは，random sampling -> data augmentation (random dilation, rotation, and cropping)
- (やらない) edge or no-edge mask と maskをhole sizeで分類する
- [x] batch normalizationのmean varianceをhole pixelに対して計算するために，initial trainingとfine-tuning phaseで学習を分割して行う．(fine-tuningの方では，bnはfreezeさせる
- [x] 現状のコードだとxavier init weightがpartial convに対して適用されていないので，そこを修正する

**Comment by tanimutomo** [comment]
model initializationをxavier initでできていることを確認した

**Comment by tanimutomo** [comment]
fine-tuningの方法としては，Inoueさんの実装を見る限りだと，500,000 iteration (initial training) -> 500,000 iteration (fine-tuned)だったので，training procedureは同じような方法でいく．とりあえず．

**Comment by tanimutomo** [comment]
torchvision.utilsのmake_gridや，save_imageはdataの型や正規化されているかなどは，無視して処理をしてくれるので，tensorのbatchをそのままinputするだけで，いい感じにvisualizeしてくれる

**Comment by tanimutomo** [comment]
inoueさんの実装を基にして，evaluateとfinetuneのためのconfigやfunctionを定義した
また，evaluateで作成された画像はlog_imageでcomet-mlに送られるように設定した

**Comment by tanimutomo** [comment]
現時点では，len(dataset) == 0なのでエラーで動かないが，そこまでは正常に動作することを確認した
あとは，
- [x] mask generatorでmaskを作成して，それを保存する
- [x] maskに関しては，data augmentationを適用するようにtransformを変更する

**Comment by tanimutomo** [comment]
- kerasの実装を基にして，MaskGeneratorで任意のマスクを作成するクラスを実装した
- data/maskに作成したマスクを保存して，それを逐次読み込んで，maskに対してはrandomresizedcropのみを設定した．
- RandomRotationはoutside pixelが0で設定されてしまうのが面倒だったので，今回は適用しないようにした

**Comment by tanimutomo** [comment]
- Done gpu-dockerのsetup
- chascaで起こったバグを修正した
よくわからないバグが行ったのと，datasetを読み込むのにかなりの時間がかかるので，localで小さなデータセットを作成して，そこで一旦実験をしてバグを修正する

**Comment by tanimutomo** [comment]
tvlossはやはりmeanが正しいっぽいので，そこは変更する．
また，tvlossの際に，逆側にずらした値も計算する必要があるので，それも含める

**Comment by tanimutomo** [comment]
- ckptのdirはtimeにあわせて作成するようにした
- tv_lossを計算するコードを，inoueさんのrepoのissueに上がっていたものを追加して，sumかmeanでどちらかを選択できるようにした
- bnをconfigのfinetuneにあわせてfreeezeするようにした
- あとは，バグ系を直した

**Comment by tanimutomo** [comment]
ベースの実装はこれで終わりなので，あとはexperiment branchで微調整を行う
